Zyphra Mamba2 and Flash Attention backward kernel blog (#25)
ROCm/rocm-blogs: blogs/artificial-intelligence/mamba2-flash-attention-kernels/README.md

Reviewer: One small thing, hopefully it doesn't take too long is to change the directory to ecosystems and partners directory. I want to showcase this as a partner contribution. Plus, this way this will stay on the front page longer as we have fewer ecosystem and partner posts.

Reviewer 2: moved

---
blogpost: true
date: 07 December 2024
author: Quentin Anthony, Beren Millidge
tags: LLM, AI/ML, GenAI, PyTorch
category: Applications & models
language: English
myst:
  html_meta:
    "description lang=en": "Mamba2 and Flash Attention Backward Kernels on AMD MI300x with ROCm"
    "keywords": "Mamba, PyTorch, S4, S6, Mamba2, Transformer, Flash Attention, Optimization, Hardware-aware, Transformer, Attention, ROCm, Mi210, MI250, MI300, AI/ML, Generative AI"
    "property=og:locale": "en_US"
---

# Training Transformers and Hybrid models on AMD Instinct MI300X Accelerators

> 

Reviewer: @Danny213123 can you help add a formatted author line here that highlights the author is from Zyphra

Author: Added a line -- let me know how I should format this officially

Reviewer 3: Hey @BerenMillidge , there is a new process of generating these author lines, it will automatically take whatever is in the metadata for authors and generate a line. If you want, you can create author pages for yourself and the other blog author, and the author line will link to those pages. You can take a look at the other authors and their bios to see what you would want to add.

On paper, the AMD Instinct MI300X GPU accelerators contain some of the best hardware specifications on the market, significantly outperforming its competitor the Nvidia H100 GPU. The key hardware specs where the MI300X surpasses the H100 are High Bandwidth Memory (HBM) capacity and bandwidth.

The MI300X also has significantly more compute hardware at its disposal, with a significantly greater number of streaming multiprocessors (SMs) than the H100. While this leads to incredibly high theoretical BFLOAT16 throughput, there are some caveats in practice that we discuss below.

For exact comparisons, see the table below:

|     | **AMD MI300X** | **NVIDIA H100 SXM** |
| --- | --- | --- |
| **Total Board Power (TBP)** | 750W | 700W |
| --- | --- | --- |
| **HBM3 Quantity** | 192GB | 80GB |
| --- | --- | --- |
| **Memory Bandwidth** | 5.3TB/s | 3.3TB/s |
| --- | --- | --- |
| **SM Count** | 304 | 113 |
| --- | --- | --- |

Despite these impressive specs, however, the Nvidia H100/H200 remains widely adopted for large-scale pretraining runs, (except Google and Apple<sup><sup>[\[1\]](#footnote-0)</sup></sup>, who use TPUs).

> This is primarily because, while the MI300X hardware is theoretically very powerful, actually materializing that performance in practice needs additional work, due to the relative nascencyof the surrounding software stack compared to NVIDIA’s CUDA ecosystem. The needfor any potential MI300X user is to pull this theoretical performance out of the underlying hardware, since many of the kernels inherent to the modern AI ecosystem were built and optimized with implicit (or explicit) assumptions around NVIDIA hardware.

Reviewer: ```suggestion
This is primarily because, while the MI300X hardware is theoretically very powerful, actually materializing that performance in practice needs additional work, due to the relative nascency of the surrounding software stack compared to NVIDIA’s CUDA ecosystem. The need for any potential MI300X user is to pull this theoretical performance out of the underlying hardware, since many of the kernels inherent to the modern AI ecosystem were built and optimized with implicit (or explicit) assumptions around NVIDIA hardware.
```

> What this means in practice is that many kernels for core operations required in pretraining either do not exist, or are poorly optimized compared to their H100 counterparts, and this negates the fundamental performance advantages of the MI300x hardware.

Reviewer: ```suggestion
What this means in practice is that many kernels for core operations required in pretraining either do not exist, or are not optimized compared to their H100 counterparts, and this negates the fundamental performance advantages of the MI300x hardware.
```

## Acknowledgements

Zyphra would like to thank TensorWave, Cirrascale, and AMD for providing us with access to AMD Instinct MI300X GPU accelerators to carry out the optimizations in this work.

1. <https://arxiv.org/abs/2407.21075> [↑](#footnote-ref-0)

2. HIP is the GPU programming language equivalent to CUDA used on AMD GPUs [↑](#footnote-ref-1)

3. Zamba2-[7B](https://www.zyphra.com/post/zamba2-7b), [2.7B](https://www.zyphra.com/post/zamba2-small), and [1.2B](https://www.zyphra.com/post/zamba2-mini) [↑](#footnote-ref-2)

4. See <https://arxiv.org/abs/2401.14489> for explanations and examples of this effect [↑](#footnote-ref-3)

5. We believe this is due to power-management firmware, which will improve over time and already has since we started working with MI300X. See <https://fireworks.ai/blog/fireattention-v3> and <https://github.com/stas00/ml-engineering/blob/master/compute/accelerator/README.md#maximum-achievable-matmul-flops-comparison-table> for further discussion. [↑](#footnote-ref-4)

> 

Reviewer: ```suggestion
```

---
blogpost: true

> date: 07 December 2024

Reviewer 4: The blog currently does not have a metadata title, using the tentative suggested title. Please paste any final title here: 
```suggestion
date: 07 December 2024
blog_title: "Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators"
```

Reviewer 5: Final title:  "Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators"

author: Quentin Anthony, Beren Millidge
tags: LLM, AI/ML, GenAI, PyTorch
category: Applications & models
language: English
myst:
  html_meta:

>     "description lang=en": "Mamba2 and Flash Attention Backward Kernels on AMD MI300x with ROCm"

Reviewer 4: "description lang=en": "This field should include a ~150 characters summary of your blog. This brief abstract-like text will appear as your blog post snippet text on the ROCm Blogs website. "

Reviewer 5: In this blog, we demonstrate the first backwards kernels to surpass H100s for both transformers (Flash Attention v2) and hybrid models (Mamba2), which enables training foundation models on AMD Instinct MI300X accelerators.

Reviewer 2: Resolved.

>     "description lang=en": "Mamba2 and Flash Attention Backward Kernels on AMD MI300x with ROCm"

Reviewer 4: See my suggestion for the blog snippet text:
```suggestion
    "description lang=en": "This blog presents Zyphra’s vision of training transformers and hybrid models at a lower cost, and its realization by utilizing and optimizing the superior hardware specs of the MI300x."
```

Reviewer 2: This was handled and is now outdated

>     "keywords": "Mamba, PyTorch, S4, S6, Mamba2, Transformer, Flash Attention, Optimization, Hardware-aware, Transformer, Attention, ROCm, Mi210, MI250, MI300, AI/ML, Generative AI"

Reviewer 4: Please add a thumbnail image, format:
thumbnail: ' image name.jpg'

Reviewer 2: Will add this today

Reviewer 2: I added the path in https://github.com/ROCm/rocm-blogs/pull/25/commits/327f685db9b9d40d2399978dabfd4fa50e78a180 and we sent the thumbnail over email. Does anything else need done for the thumbnail?

    "property=og:locale": "en_US"
---

> # Training Transformers and Hybrid models on AMD Instinct MI300X Accelerators

Reviewer 4: Let me know what you think of this suggested tentative title (see metadata below).
Other suggestions are:
**Diamond in the Rough: Zyphra Speeding Up Training on AMD Instinct MI300X Accelerators**
Or simply:
**Zyphra Speeding Up Training on AMD Instinct MI300X Accelerators**
```suggestion
# Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators
```

Reviewer 5: Title shared above "Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators"

> *By Quentin Anthony and Beren Millidge from Zyphra*

Reviewer 4: Please see my suggested text in the new intro section. Please also see @Danny213123 earlier suggestion to add brief bios to ROCm Blogs 
```suggestion

```

> 

Reviewer 4: We are following a magazine article approach where there is no explicit introduction perse, but rather the blog starts with a brief, wide-scoped introductory text before moving into the blog’s first section

```suggestion
This blog is contributed by [Zyphra](https://www.zyphra.com/): a Palo Alto based AI research lab and AMD Instinct Partner.

Zyphra is designing MaiaOS, a multimodal agent system that combines next-gen neural network architectures (SSM hybrids), long-term memory & reinforcement learning.

In this blog we motivate our vision of training transformers and hybrid models at a lower cost using AMD technology. We explain how Zyphra harnessed the fundamental performance advantages of the MI300x hardware on Zyphra's hybrid models: the Mamba2, and the Flash Attention v2. We conclude the blog by sharing benchmarks results showing the speedups we achieved on the MI300X using ROCm, compared to the competition.
```

> On paper, the AMD Instinct MI300X GPU accelerators contain some of the best hardware specifications on the market, significantly outperforming its competitor the Nvidia H100 GPU. The key hardware specs where the MI300X surpasses the H100 are High Bandwidth Memory (HBM) capacity and bandwidth.

Reviewer 4: ```suggestion
## Harnessing the MI300 Superior Hardware Specs

On paper, the AMD Instinct MI300X GPU accelerators contain some of the best hardware specifications on the market, significantly outperforming its competitor the Nvidia H100 GPU. The key hardware specs where the MI300X surpasses the H100 are High Bandwidth Memory (HBM) capacity and bandwidth.
```

Reviewer 2: moved

Reviewer 2: I created a subsection with the more neutral wording of `## Introduction`

![png](./images/Mamba2_kernel_backward_AMD_blog.png)

We baked in the following AMD-specific optimizations when writing the Mamba-2 backward kernel:

1. Improved shared memory movement and access by arranging the smem_delta_a, smem_running_postfix, smem_da, smem_dbc buffers to better align with MI300X hardware, and within 64KB.
2. Modified prefix scan and reverse scan operations to the MI300X wavefront size. Perform tuning of block sizes and kernel launch parameters similar to FA2 to be optimal on MI300X.
3. Rework the block-exchange for atomicAdds using patterns more amenable to MI300X while maintaining coalescing.

Similar to FA2, we achieve speedups on the Mamba2 backward kernel of 4%, 5%, and 6% for sequence lengths of 2k, 4k, and 8k, respectively on MI300X compared to the H100. Cache thrashing, data movement cost, and SM utilization are all significantly improved. With both the Mamba2 forwards and backwards and the Flash Attention 2 backwards kernel in-hand, pure-SSM and hybrid attention/SSM models are trainable on MI300X hardware, and can achieve higher FLOPs per dollar than is possible on NVIDIA H100 systems.

> ## Future Work

Reviewer 4: Adding a brief summary, which includes the future work section:
```suggestion
## Summary

In this blog post we outlined Zyphra’s vision of training transformers and hybrid models at a lower cost. We explain how Zyphra is realizing this vision by optimizing the superior hardware specifications of the AMD Instinct MI300X Accelerators, using ROCm to train Zyphra’s hybrid models: the Mamba2 and the Flash Attention v2.

As future work, Zyphra plans to extend the attention kernel and portions of the Mamba2 kernel to fp8 precision, and enable fine-grained tensor-parallel overlap within the Mamba2, Attention, and MLP blocks with communication. Both optimizations are critical to Zyphra’s training pipeline.

```

---
blogpost: true
date: 07 December 2024
blog_title: "Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators"
author: Quentin Anthony, Beren Millidge
tags: LLM, AI/ML, GenAI, PyTorch
category: Applications & models
language: English
myst:
  html_meta:

>     "description lang=en": "Mamba2 and Flash Attention Backward Kernels on AMD MI300x with ROCm"

Reviewer: ```suggestion
    "description lang=en": "In this blog, we demonstrate the first backwards kernels to surpass H100s for both transformers (Flash Attention v2) and hybrid models (Mamba2), which enables training foundation models on AMD Instinct MI300X accelerators."
```

---
blogpost: true
date: 07 December 2024
blog_title: "Zyphra Introduces Frontier Training Kernels for Transformers and SSMs on AMD Instinct MI300X Accelerators"
author: Quentin Anthony, Beren Millidge
tags: LLM, AI/ML, GenAI, PyTorch

> category: Applications & models

Reviewer: ```suggestion
category: Ecosystems and Partners
```

Similar to FA2, we achieve speedups on the Mamba2 backward kernel of 4%, 5%, and 6% for sequence lengths of 2k, 4k, and 8k, respectively on MI300X compared to the H100. Cache thrashing, data movement cost, and SM utilization are all significantly improved. With both the Mamba2 forwards and backwards and the Flash Attention 2 backwards kernel in-hand, pure-SSM and hybrid attention/SSM models are trainable on MI300X hardware, and can achieve higher FLOPs per dollar than is possible on NVIDIA H100 systems.

## Summary

In this blog post we outlined Zyphra’s vision of training transformers and hybrid models at a lower cost. We explain how Zyphra is realizing this vision by optimizing the superior hardware specifications of the AMD Instinct MI300X Accelerators, using ROCm to train Zyphra’s hybrid models: the Mamba2 and the Flash Attention v2.

As future work, Zyphra plans to extend the attention kernel and portions of the Mamba2 kernel to fp8 precision, and enable fine-grained tensor-parallel overlap within the Mamba2, Attention, and MLP blocks with communication. Both optimizations are critical to Zyphra’s training pipeline.

> 

Reviewer: ```suggestion
```
Fix linting
